### DISCUSSION TITLE
Blazor MAUI

### DISCUSSION THREAD

**Discussion by Sen-Gupta** [oqtane/oqtane.framework#2347] [Ideas] [answered]
Hello @sbwalker 

I see a Blazor MAUI implmentation for this planned for 
![image](https://user-images.githubusercontent.com/8068582/183401818-19022ca6-7f6c-4b9d-9d14-8f0eb3075cf0.png)

I was wondering if it would support Dynamic Module/dll download on client side as we see currently.

I think, it may be possible for for Blazor MAUI, if at all?

But for Natives, do you think, it is possible?

I really do not  have much clarity on what features will be supported in Blazor MAUI vs Native developed on MAUI?

Would you mind shedding some light and ideas?

**Comment by sbwalker** [comment]
The MAUI integration will be focused on Blazor Hybrid - which essentially uses an embedded WebView to render components natively. Blazor Hybrid uses a strict client/server model so the plan is to use a similar approach as what is being used for Blazor WebAssembly to transfer and load assemblies dynamically into the app domain.

**Reply by Sen-Gupta** [reply]
@sbwalker thanks for the response.

What I hear from above, that "**Yes**" it would be possible to load  module assemblies dynamically! ;-) in case of Blazor MAUI.

I have been  going through whatever little samples avaiable for Blazor MAUI and everywhere, the Pages are kept in a separate assembly like MyProject.Pages and then those pages are referred in both Blazor MAUI and WebAssembley and/or Blazor Server. 

Which means, we need to Oqtane to follow that? and what about each module bringing in there own pages?

I do believe that  abve two cases are solvable but we don't have enough sample solution for lazy loading assemblies with Blazor MAUI.
We have been using Lazy Loading techniques of Microsoft and also studied the ones implemented in Oqtane, I like oqtane better as if there is a module which is secured and must be downloaded 
This is my grey area!

**Reply by sbwalker** [reply]
@Sen-Gupta I believe that it is technically possible to dynamically load module assemblies in Blazor Hybrid, however I have not proven it yet. My initial focus is to get the framework to run on Blazor Hybrid and once that is successful I plan to focus on the extensions. One item to note is that Oqtane does not use "pages"... in Blazor a "page" is a component with an @page directive which is usually located in a Pages folder for convenience. These are also called "routable components" and Oqtane does not support them because it uses its own custom router - not the default Blazor router.

**Comment by sbwalker** [comment, marked as answer]
Making progress...

![image](https://user-images.githubusercontent.com/4840590/183760698-30feedcd-7470-4b98-9b70-a3e3f147a9f0.png)

Discovering a lot of small challenges along the way... mostly related to the fact that a Blazor application running within a BlazorWebView does not handle relative Urls in the same manner as it does when running on Blazor WebAssembly. Blazor WebAssembly will map relative Urls to the server using the current browser domain.... Blazor Hybrid assumes the relative Urls refer to local resources on the client.

**Comment by Sen-Gupta** [comment]
@sbwalker Awesome to see it running already!

Perhaps, I would like to add where I'm coming from on this!

Developing a muliplatform app is one thing, and get them through the doors of App Stores altogather another series of tasks. There are various policies applicable, of course all for security of people and platform but this process of getting approvals from App Store everytime for updates of the Application is tedious and poor Go to Market.

Now Imagine this:
Oqtane based app downloaded as Native App* (Running on Blazor Web View)  from App Stores onces and Voila all updates to app can be delivered dynamically as the updated dlls can be pulled from App automatically so there is never a need to update apps and go through App stores tedious process.
On the other hand for customers they are always on updated apps automagically!

That's the value propositon that your implementation above brings atleast to my understanding!

With this Oqtane is really going to be on "Rocket Fuel".

Am I thinking too much?

By the way, the sample that you above show-cased is avaialble on the Github? or it is in private repo ;-)

**Reply by sbwalker** [reply]
No you are not thinking too much... this dynamic servicing of an app is fundamental to the Oqtane approach and Maui opens up exciting new opportunities. 

The code is still a work in progress locally - not ready for merging into the dev branch yet. I need to work through a few more issues first.

**Comment by sbwalker** [comment]
More progress...

![image](https://user-images.githubusercontent.com/4840590/184020630-d8ebc3ca-8d05-4502-9b25-b1686cd99237.png)

**Comment by sbwalker** [comment]
#2355 includes the changes required for Oqtane run on .NET MAUI. 

There is a new Oqtane.Maui.sln file in the root folder (the Oqtane.Maui project currently has a dependency on the Oqtane.Client and Oqtane.Shared assemblies in the /bin folder - so you would need to load Oqtane.sln and build those projects before opening the Oqtane.Maui solution). 

In order to run the Maui application you need to have the Oqtane.Server running and the easiest way to do this is to configure your startup projects in Visual Studio so that both Oqtane.Server and Oqtane.Maui will startup when you run the application:

![image](https://user-images.githubusercontent.com/4840590/184244126-6a18ddac-2e3a-4d81-ad95-af505a564765.png)

For simplicity I have been testing the Maui app in Visual Studio using the Windows Machine option, as it does not require the installation or configuration of an emulator. However you do need to be sure to enable Developer Mode in Windows:

![image](https://user-images.githubusercontent.com/4840590/184244550-791fac8d-3213-4d8b-b0dc-9cb6a9626b12.png)

Once Oqtane is running, you can login and use all of the same functionality that is available on Blazor Server or Blazor WebAssembly. In addition, the framework still supports run-time deployment of modules and themes ( as evidenced in the screenshot below where the Demo31 external module is running in Maui ): 

![image](https://user-images.githubusercontent.com/4840590/184243148-742945c4-ffe5-444b-a2f4-d6a54857713c.png)

More testing is definitely required but for the most part the framework is fully functional on this new platform.
